Rubberduck.Parsing/VBA/ParseCoordinator.cs

```diff
@@ -86,7 +103,80 @@ private void ReparseRequested(object sender, EventArgs e)
                 token = _currentCancellationTokenSource.Token;
             }
 
+            BeginParse(sender,token);
+        }
+
+        public void SuspendRequested(object sender, RubberduckStatusSuspendParserEventArgs e)
+        {
+            object parseRequestor = null;
+            ParserState originalStatus;
 
+            try
+            {
+                if (_parsingSuspendLock.IsReadLockHeld)
+                {
+                    const string errorMessage =
+                        "A suspension action was attempted while a read lock was held. This indicates a bug in the code logic as suspension should not be requested from same thread that has a read lock.";
+                    Logger.Error(errorMessage);
+#if DEBUG
+                    Debug.Assert(false, errorMessage);
+#endif
+                    return;
+                }
+                if (!_parsingSuspendLock.TryEnterWriteLock(e.MillisecondsTimeout))
+                {
+                    e.TimedOut = true;
+                    return;
+                }
+
+                lock (_suspendStackSyncObject)
+                {
+                    _isSuspended = true;
+                }
+
+                originalStatus = State.Status;
+                _parserStateManager.SetStatusAndFireStateChanged(e.Requestor, ParserState.Busy,
+                    CancellationToken.None);
+                e.BusyAction.Invoke();
+
+                lock (_suspendStackSyncObject)
+                {
+                    _isSuspended = false;
+                    if (_requestorStack.TryPop(out var lastRequestor))
+                    {
+                        _requestorStack.Clear();
+                        parseRequestor = lastRequestor;
```

Comment:
I'm guessing you meant passing `parseRequestor` -- primarily because I don't like to dirty a variable from a `out`/`ref` where there has to be a check on the state. The documentation for `TryOut` does not guarantee that the object will be `null`, merely "unspecified". This code guarantees that there will be no side effect on the `parseRequestor` since we later check whether it's non-null to determine if we must reparse or not.